File: LeetCode - Top Interview 150/82.py
# 82. Remove Duplicates from Sorted List II
# https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/


from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    # time complexity: O(n)
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        start_node = ListNode(-500, head)

        if not head or not head.next:
            return head

        first = start_node
        second = head
        third = head.next

        while first and second and third:
            if second.val == third.val:
                while third and second.val == third.val:
                    third = third.next

                if not third:
                    first.next = None
                else:
                    first.next = third
                    second = first.next
                    third = second.next

                continue

            first = first.next
            second = second.next
            third = third.next

        return start_node.next

File: LeetCode - Algorithm/Linked List/82.js
// 82. Remove Duplicates from Sorted List II
// https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function (head) {
  if (!head) return head;

  const dummy = new ListNode(-1);
  let current = dummy;

  let slow = head;
  let fast = head;

  while (fast.next) {
    if (fast.val === fast.next.val) {
      fast = fast.next;
    } else if (slow === fast) {
      current.next = fast;
      current = current.next;
      current.next = null;

      slow = fast.next;
      fast = fast.next;

      current.next = null;
    } else {
      slow = fast.next;
      fast = fast.next;
    }
  }

  if (slow === fast) {
    current.next = fast;
    current = current.next;
  }

  return dummy.next;
};

/*
head = [1,2,3,3,4,4,5]
                    ^

head = [1,1,1,2,3]
                ^

head = [1,2,2]
          ^ ^
*/
